torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,289 @@
+        lmbda (float, optional): the lambda discount factor factor. default: 0.95
+        discount_loss (bool, optional): if True, the loss is discounted with a gamma discount factor. default: True.
+    """
+
+    def __init__(
+        self,
+        actor_model: TensorDictModule,
+        value_model: TensorDictModule,
+        model_based_env: DreamerEnv,
+        imagination_horizon: int = 15,
+        gamma: int = 0.99,
+        lmbda: int = 0.95,
+        discount_loss: bool = True,
+    ):
+        super().__init__()
+        self.actor_model = actor_model
+        self.value_model = value_model
+        self.model_based_env = model_based_env
+        self.imagination_horizon = imagination_horizon
+        self.gamma = gamma
+        self.lmbda = lmbda
+        self.discount_loss = discount_loss
+
+    def forward(self, tensordict: TensorDict) -> torch.Tensor:
+        with torch.no_grad():
+            tensordict = tensordict.select("posterior_state", "belief", "reward")
+
+            tensordict.batch_size = [
+                tensordict.shape[0],
+                tensordict.get("belief").shape[1],
+            ]
+            tensordict.rename_key("posterior_state", "prior_state")
+            tensordict = tensordict.view(-1).detach()
+        with hold_out_net(self.model_based_env), set_exploration_mode("random"):
+            tensordict = self.model_based_env.rollout(
+                max_steps=self.imagination_horizon,
+                policy=self.actor_model,
+                auto_reset=False,
+                tensordict=tensordict,
+            )
+            tensordict = step_tensordict(
+                tensordict,
+                keep_other=True,
+                exclude_reward=False,
+                exclude_action=False,
+            )
+            with hold_out_net(self.value_model):
+                tensordict = self.value_model(tensordict)
+
+        lambda_target = self.lambda_target(
+            tensordict.get("reward"), tensordict.get("predicted_value")
+        )
+        tensordict = tensordict[:, :-1]
+        tensordict.set("lambda_target", lambda_target)
+
+        if self.discount_loss:
+            discount = self.gamma * torch.ones_like(
+                lambda_target, device=tensordict.device
+            )
+            discount[:, 0] = 1
+            discount = discount.cumprod(dim=1).detach()
+            actor_loss = -(lambda_target * discount).mean()
+        else:
+            actor_loss = -lambda_target.mean()
+        return (
+            TensorDict(
+                {
+                    "loss_actor": actor_loss,
+                },
+                batch_size=[],
+            ),
+            tensordict.detach(),
+        )
+
+    def lambda_target(self, reward: torch.Tensor, value: torch.Tensor) -> torch.Tensor:
+        done = torch.zeros(reward.shape, dtype=torch.bool, device=reward.device)
+        return vec_td_lambda_return_estimate(
```

Comment:
(3) now we take the values from the second and on (hence the next_next_state) + we discard one datapoint)
I think it's wrong, we should just take value and reward and done, without indexing them